Hax extraction and F* lax-checking

maximebuyse opened the pull request:
This PR replaces https://github.com/cryspen/formats/pull/1, with the changes described in https://github.com/cryspen/home/issues/399#issuecomment-3160178752.
It implements the necessary workarounds (and adds some infrastructure) to have working hax extraction + F* lax-checking. Extraction can be done with `cargo hax into fstar`, and lax-checking with `OTHERFLAGS="--admit_smt_queries true" make -C proofs/fstar/extraction/`.

We have workarounds for:
- https://github.com/cryspen/hax/issues/1595
- https://github.com/cryspen/hax/issues/888
- Some unsupported mutability
- https://github.com/cryspen/hax/issues/1512
- https://github.com/cryspen/hax/issues/1584

maximebuyse committed:
Replace 'vec_mut' for compatibility with hax capacities
.

maximebuyse committed:
Hax workaround for default in traits.

maximebuyse committed:
Hax workaround for method name clash.

maximebuyse committed:
Various hax workarounds.

maximebuyse committed:
Avoid bundle for hax method names clash workaround.

maximebuyse committed:
Hax infrastructure and working lax-checking.

franziskuskiefer commented on tls_codec/proofs/fstar/extraction/.gitignore:
```diff
@@ -0,0 +1,4 @@
+*
+
+!.gitignore
+!Makefile
```
```suggestion
!Makefile

```

franziskuskiefer commented on tls_codec/src/quic_vec.rs:
```diff
@@ -127,8 +127,9 @@ pub fn write_variable_length(content_length: usize) -> Result<Vec<u8>, Error> {
     let mut len = content_length;
-    for b in length_bytes.iter_mut().rev() {
-        *b |= (len & 0xFF) as u8;
+    let l = length_bytes.len();
```
You listed the hax issues on the PR. Can you add them to each of the changes as well? Then it's easier to understand why it has to be this way, and we can find all places later when we fix the hax issues.

maximebuyse replied:
Done (comments in the code).

franziskuskiefer commented on tls_codec/src/quic_vec.rs:
```diff
@@ -251,18 +254,6 @@ macro_rules! impl_vl_bytes_generic {
-
-            /// Add an element to this.
-            #[inline]
-            pub fn push(&mut self, value: u8) {
```
These are public functions. We should keep them. Deleting them would be a breaking change. Let's just ignore them for hax extraction for now.

franziskuskiefer commented on tls_codec/src/quic_vec.rs:
```diff
@@ -311,8 +302,16 @@ impl VLBytes {
-        &mut self.vec
+    /// Add an element to this.
+    #[inline]
+    fn push(&mut self, v: u8) {
```
The compiler says this is dead code. Also, isn't this the function that you deleted from the macro above? I'm not sure I understand this change.

maximebuyse replied:
Yes, this is intended to replace the function from the macro. I don't understand why we get the dead code warning in this case, and not with the macro. And this shouldn't be a breaking change as the functions are replaced. But we can also revert and ignore them for hax extraction.

franziskuskiefer replied:
Maybe this function should be `pub`?

franziskuskiefer commented on tls_codec/Cargo.toml:
```diff
@@ -16,6 +16,7 @@ zeroize = { version = "1.8", default-features = false, features = [
     "alloc",
     "zeroize_derive",
 ] }
+hax-lib = {git = "https://github.com/cryspen/hax", branch = "tls-codec-lib"}
```
We should track somewhere to move to a released version here.

maximebuyse replied:
The lib changes are merged into evit, and part of the next approval batch. So once we get the approval we can upstream and make a release. I can open an issue to track.

franziskuskiefer commented on tls_codec/src/lib.rs:
```diff
@@ -154,13 +168,37 @@ pub trait Serialize: Size {
-/// * `tls_serialize` that returns a byte vector
-pub trait SerializeBytes: Size {
+#[cfg(hax)]
+pub trait SerializeDetached {
```
Shall we move these into a `hax` module to make it a little easier to read?

maximebuyse replied:
Done

franziskuskiefer commented on tls_codec/src/lib.rs:
```diff
@@ -198,6 +236,41 @@ pub trait Deserialize: Size {
+    where
+        Self: Sized,
+    {
+        let mut bytes = bytes.as_ref();
```
We can move out the actual code to a standalone function to avoid duplicating code. Then the extraction would also have the actual code that's being run.

maximebuyse replied:
Done

franziskuskiefer commented on tls_codec/derive/tests/decode.rs:
```diff
@@ -3,6 +3,8 @@ use tls_codec::{
     Deserialize, DeserializeBytes, Error, Serialize, Size, TlsSliceU16, TlsVecU8, TlsVecU16,
     TlsVecU32, VLBytes,
 };
+#[cfg(hax)]
```
The derive crate needs the `check-cfg` for `hax` as well to avoid the warnings here.

maximebuyse replied:
Done

franziskuskiefer reviewed (commented):
Looks pretty good already. A few comments.

maximebuyse committed:
Update tls_codec/proofs/fstar/extraction/.gitignore

Co-authored-by: Franziskus Kiefer <franziskuskiefer@gmail.com>

maximebuyse committed:
Changes for review feedback.

franziskuskiefer reviewed (commented):
It looks like the CI fails for msrv and no_std. Can you double check what's going on there? If you don't understand why it's broken and think something in the check is bad, let me know and I can have a look. but from the error, it looks there's a breaking change that makes the CI fail.

maximebuyse committed:
Small fixes for CI.

maximebuyse committed:
Fix no-default-feature with hax workarounds.

maximebuyse committed:
Add hax module for default trait methods workaround.

maximebuyse committed:
Add hax extraction and f* lax-checking on CI.

maximebuyse commented:
> It looks like the CI fails for msrv and no_std. Can you double check what's going on there? If you don't understand why it's broken and think something in the check is bad, let me know and I can have a look. but from the error, it looks there's a breaking change that makes the CI fail.

I fixed that. I also added a CI job for hax extraction + lax-checking.

maximebuyse committed:
Fix warnings under hax feature.

maximebuyse committed:
Fix more hax feature warnings.

maximebuyse committed:
Fix hax CI job for tls_codec.

franziskuskiefer reviewed (commented):
Thanks! I think this looks pretty good now.

I'd be happy to merge this here, or upstream. Depending on whether you want to add more for a first round.

maximebuyse commented:
> Thanks! I think this looks pretty good now.
> 
> I'd be happy to merge this here, or upstream. Depending on whether you want to add more for a first round.

Thanks! I think this is a good first step, let's merge. To merge upstream, maybe it is better to wait for the hax-lib changes to be released.

franziskuskiefer reviewed (approved):
Sounds good. Let's get it in then
